Refactor channel_group_search implementation in tdm_loader

EspenEnes opened the pull request:
When searching and resulting groups has identical channel group names, only the first channel group index is returned

New logic will look at channel group names together with occurrences to find the correct channel group index.

espenenesNOV committed:
Refactor channel_group_search implementation in tdm_loader

When searching and resulting groups has identical channel group names, only the first channel group index is returned

New logic will look at channel group names together with occurrences to find the correct channel group index.

domna commented on tdm_loader/tdm_loader.py:
```diff
@@ -185,22 +185,27 @@ def channel_group_search(self, search_term):
+                    group_names[x.text] = 0
+                else:
+                    group_names[x.text] += 1
+                chg_names.append((f"{x.text}", group_names[x.text]))
```
This can just be `x.text`, no?
```suggestion
                chg_names.append((x.text, group_names[x.text]))
```

domna commented on tdm_loader/tdm_loader.py:
```diff
@@ -185,22 +185,27 @@ def channel_group_search(self, search_term):
         for name in found_terms:
-            i = chg_names.index(name)
-            ind.append((name, i))
+            i = self.channel_group_index(name[0], name[1])
+            ind.append((name[0], i))
```
Couldn't you just do a simple for-loop here and omit the code above? (Maybe I'm overlooking something)

```suggestion
            for occurence in range(found_terms.count(name)):
               i = self.channel_group_index(name, occurence)
               ind.append((name, i))
```

EspenEnes replied:
Yes, the simplest is the best, there is a lot of roads to Rome and mine was the long way around  :)
I tested the simplified  suggestion and gave me the results I wanted.

domna reviewed (commented):
Thank you for your contribution! Here are some comments on a slightly simpler implementation (but maybe I'm overlooking something here).

Do you have an example file with which we could add a test for this?

EspenEnes committed:
Update tdm_loader/tdm_loader.py

Co-authored-by: Florian Dobener <github@schroedingerscat.org>

espenenesNOV committed:
Simplify channel group search logic in tdm_loader

The previous implementation counted and used the number of times a channel group name appeared, which added extra complexity to the search logic. The new logic simplifies this process by directly appending the names to the found_terms list and by making the search case- and space-insensitive.

espenenesNOV committed:
Refactor search logic in tdm_loader for channel group names

Switched from a for-loop and conditional check structure to an inline list comprehension to search for channel group names. The aim of this refactor is to simplify the code and improve readability, while maintaining the functionality of ignoring None values and retaining valid group names.

EspenEnes commented:
The latest commit I have used yours suggestion and removed my implementation to find the occurrences with the for loop.

espenenesNOV committed:
Update return type in tdm_loader

Changed the return type of a function in tdm_loader from a list to a set. This modification ensures unique channel group names are returned from the function, thus improving the accuracy of the channel search process.

domna commented:
Thank you for the changes. One of the tests are not passing now, because you changed the return type of the function from list to set. I think it would also be useful to not use a set here as people might expect that this function returns a list. I would suggest that you do a check whether a certain key already exists before inputting it into the list (or you use a set in the for-loop and convert it to a list in the return but you might need to sort it to pass the test again).

Do you also have a file containing duplicate names which we could add as a test for your new feature?

domna commented on tdm_loader/tdm_loader.py:
```diff
@@ -199,10 +200,11 @@ def channel_group_search(self, search_term):
-            i = chg_names.index(name)
-            ind.append((name, i))
+            for occurence in range(found_terms.count(name)):
+               i = self.channel_group_index(name, occurence)
+               ind.append((name, i))
```
There is one space to less (probably from me when you just accepted my suggestion)

```suggestion
                i = self.channel_group_index(name, occurence)
                ind.append((name, i))
```

espenenesNOV committed:
Removed the use of set when returning search results from channel_group_search

EspenEnes commented:
**Sorry for my clumsiness, this is my first contribution ever.**
I removed the set and used a dictionary lookup instead for the occurrences.
I have run your tests all passed..

I did not have any small TDM/TDX files I could share to make a test, all my files are 500mb ++ and Autogenerated from a recorder.

domna commented on tdm_loader/tdm_loader.py:
```diff
@@ -198,9 +199,13 @@ def channel_group_search(self, search_term):
         for name in found_terms:
-            i = chg_names.index(name)
-            ind.append((name, i))
+            if name not in occurences:
+                occurences[name] = True
+                for occurence in range(found_terms.count(name)):
+                   i = self.channel_group_index(name, occurence)
+                   ind.append((name, i))
```
```suggestion
            for occurence in range(found_terms.count(name)):
                i = self.channel_group_index(name, occurence)
                if (name, i) not in ind:
                    ind.append((name, i))
```

domna replied:
This can also be simpler

domna commented on tdm_loader/tdm_loader.py:
```diff
@@ -190,6 +190,7 @@ def channel_group_search(self, search_term):
             for x in self._root.findall(".//tdm_channelgroup/name")
             if x.text is not None
         ]
+
```
```suggestion
```

domna commented on tdm_loader/tdm_loader.py:
```diff
@@ -198,9 +199,13 @@ def channel_group_search(self, search_term):
         ]
 
         ind = []
+        occurences = {}
```
```suggestion
```

domna reviewed (changes requested):
I think there is also a more straightforward approach to this

domna commented:
> **Sorry for my clumsiness, this is my first contribution ever.** I removed the set and used a dictionary lookup instead for the occurrences. I have run your tests all passed..

No worries. This is totally fine and part of the review process. Thank you for the changes again. Also congrats on your first contribution.
> 
> I did not have any small TDM/TDX files I could share to make a test, all my files are 500mb ++ and Autogenerated from a recorder.

Ok this is too large unfortunately. There is not chance you are able to generate a very small testing file just containing just some duplicate axes?

EspenEnes committed:
Update tdm_loader/tdm_loader.py

Co-authored-by: Florian Dobener <github@schroedingerscat.org>

EspenEnes committed:
Update tdm_loader/tdm_loader.py

Co-authored-by: Florian Dobener <github@schroedingerscat.org>

EspenEnes committed:
Update tdm_loader/tdm_loader.py

Co-authored-by: Florian Dobener <github@schroedingerscat.org>

EspenEnes commented:
suddenly it became such an little fix :)

domna commented:
Thank you @EspenEnes. I'll merge it now and create a new release with it so you can use it
